unstructured/ingest/connector/vectara.py

```diff
@@ -0,0 +1,246 @@
+    WriteConfig,
+)
+from unstructured.ingest.logger import logger
+from unstructured.staging.base import flatten_dict
+
+
+@dataclass
+class VectaraAccessConfig(AccessConfig):
+    oauth_client_id: str = enhanced_field(sensitive=True)
+    oauth_secret: str = enhanced_field(sensitive=True)
+
+
+@dataclass
+class SimpleVectaraConfig(BaseConnectorConfig):
+    access_config: VectaraAccessConfig
+    customer_id: t.AnyStr = None
+    corpus_name: t.AnyStr = "vectara-unstructured"
+    corpus_id: t.AnyStr = None
+
+
+@dataclass
+class VectaraWriteConfig(WriteConfig):
+    batch_size: int = 100
+
+
+@dataclass
+class VectaraDestinationConnector(BaseDestinationConnector):
+    write_config: VectaraWriteConfig
+    connector_config: SimpleVectaraConfig
+
+    BASE_URL = "https://api.vectara.io/v1"
+
+    @DestinationConnectionError.wrap
+    def vectara(self):
+        """
+        Check the connection for Vectara and validate corpus exists.
+        - If more than one exists - then return a message
+        - If exactly one exists with this name - use it.
+        - If does not exist - create it.
+        """
+        try:
+            jwt_token = self._get_jwt_token()
+            if not jwt_token:
+                return "Unable to get JWT Token. Confirm your Client ID and Client Secret."
+
+            list_corpora_response = self._request(
+                endpoint="list-corpora",
+                data={"numResults": 100, "filter": self.connector_config.corpus_name},
+            )
+
+            possible_corpora_ids_names_map = {
+                corpus.get("id"): corpus.get("name")
+                for corpus in list_corpora_response.get("corpus")
+                if corpus.get("name") == self.connector_config.corpus_name
+            }
+
+            if len(possible_corpora_ids_names_map) > 1:
+                return f"Multiple Corpora exist with name {self.connector_config.corpus_name}"
+            if len(possible_corpora_ids_names_map) == 1:
+                self.connector_config.corpus_id = list(possible_corpora_ids_names_map.keys())[0]
+            else:
+                data = {
+                    "corpus": {
+                        "name": self.connector_config.corpus_name,
+                    }
+                }
+                create_corpus_response = self._request(endpoint="create-corpus", data=data)
+                self.connector_config.corpus_id = create_corpus_response.get("corpusId")
+
+        except Exception as e:
+            return str(e) + "\n" + "".join(traceback.TracebackException.from_exception(e).format())
+
+    def initialize(self):
+        self.vectara()
+
+    def _request(
+        self,
+        endpoint: str,
+        http_method: str = "POST",
+        params: t.Mapping[str, t.Any] = None,
+        data: t.Mapping[str, t.Any] = None,
+    ):
+        url = f"{self.BASE_URL}/{endpoint}"
+
+        current_ts = datetime.datetime.now().timestamp()
+        if self.jwt_token_expires_ts - current_ts <= 60:
```

Comment:
Do we ever have to worry about token refresh since the connector is fairly short lived and it'll fetch a new jwt token every time it's instantiated. Also we should set a `_jwt_token` variable as part of the dataclass defaulted to None, and maybe have a `@property` that returns a jwt token, which can be responsible for fetching a new one if the instance variable is `None`. Similar to how we cache clients.

Reply:
Great idea. Done.